sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathCollector.java

```diff
@@ -7,40 +7,77 @@
 import java.util.Map;
 
 /**
- * Captures only the ApexPath of interest, allows the rest of the ApexPathExpander objects to be
- * garbage collected.
+ * Optimization to reduce the number of objects that are kept around while expanding paths. When an
+ * {@link ApexPathExpander} is accepted, its associated {@link ApexPath} is captured. When it is
+ * rejected, the {@link PathExpansionException} that caused the rejection is captured. Either way,
+ * the rest of the {@link ApexPathExpander} can be discarded and can be garbage-collected, thus
+ * relieving memory pressure.
  */
-final class ApexPathCollector extends ResultCollector {
+public final class ApexPathCollector {
     /**
      * Keep a map of the Expander Id to the ApexPath. ApexPath prevents equality comparisons to
      * avoid bugs
      */
     private final Map<Long, ApexPath> expanderIdToPath;
+    /**
+     * Mapping from the IDs of expanders whose paths were rejected to the reasons for those
+     * rejections.
+     */
+    private final Map<Long, PathExpansionException> expanderIdToRejectionReason;
```

Comment:
We now track the reasons that paths were rejected, so we can pass them further up the chain.